rfcs/0040-rosetta-timelocking.md

```diff
@@ -0,0 +1,119 @@
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+### Balance exemptions and historical balance lookups
+
+Rosetta supports the notion of a balance exempt account. These are accounts that one should not use operations alone to monitor changes in balance. The specification details that this should be used sparingly, but goes on to suggest that vesting accounts are one such example. If we only support time-locked accounts in the genesis ledger, all we need to do is at Rosetta-server-startup-time is pull those time-locked account addresses and fill the balance exemption field. If we don't, we'll need to update this field dynamically, it remains to be seen if this is allowable by the Rosetta specification, see unresolved question 1.
+
+This is not a solution alone -- the specification goes on to say that in the case that one or more accounts are balance exempt, you must support historical balance lookups. This is difficult for us because Mina's constant-sized blockchain does not contain historical data. We use the archive node to store extra data for us -- we're already using it to store all the blocks that our nodes see. In order to support historical balance lookups, we have to add extra information so we can compute this data. Specifically, we'll add information about the current balance of any accounts touched during a transaction whenever we store in the archive node. We can use this and the genesis ledger from Rosetta to reconstruct the current balance of any account at any block by crawling backward from that block until we see a transaction (or fallback to the genesis ledger if no such transaction exists). Additionally, we can use the extra timing information in time-locked accounts in the genesis ledger to sample the liquidity curve at any moment as well. In the universe where we need to support time-locked account creation after genesis, we should be able to extract the timing information from the relevant transaction that creates the account via the archive node as well.
+
+### Implementation details
+
+In some cases, two separate approaches are outlined. This RFC proposes that we go with Approach A in the case that by mainnet we do not need to support creating time-locked accounts after the genesis ledger, and Approach B if we do end up needing to support this feature.
+
+**Protocol/Archive**
+
+- Add a new table `balances` to the SQL database with the following schema:
+
+```
+id,
+public_key,
+amount
+```
+
+- Add a new table `user_commands_balances` to the SQL database with the following schema:
+
+```
+user_command_id,
+balance_id
+```
+
+- Add a new table `internal_commands_balances` to the SQL database with the following schema:
+
+```
+internal_command_id,
+balance_id
```

Comment:
Similarly here, wrt the `blocks_internal_commands` table.

@psteckler commented on #6931
> Also, commands are not applied in a way that there's a known balance corresponding to a given command. In the replayer, commands are sometimes combined, and the balance reconstruction would have to accomodate a similar mechanism.

I think it may be incorrect to think of the internal commands as existing separate from the `blocks_internal_commands` table, even though we store them as such. IIRC the rosetta spec demands that transactions (or their equivalent concept) have a unique identifier; we would need to consider them in their context of the block -- that is, using the `block_id` and possibly `sequence_no` as part of the command identifier -- to ensure that the commands are not considered the same. This feels like good practice to me regardless.

I think also that storing the data in this table elides the concern about balances: once associated with the particular sequence number and block id, the balances are entirely determined.

Reply:
Yes, even in the case in combined internal commands, or combined user commands, the combined items always match on sequence number and, if applicable, secondary sequence number, so the balances will be the same for both of the combined items.